Spigot-Server-Patches/0406-implement-optional-per-player-mob-spawns.patch

```diff
@@ -0,0 +1,259 @@
++                                                                return mobsSpawned;
+                                                             }
++                                                            // Paper end
+                                                             // CraftBukkit end
+                                                             if (i >= entityinsentient.dC()) {
+-                                                                return;
++                                                                return mobsSpawned; // Paper
+                                                             }
+ 
+                                                             if (entityinsentient.c(i2)) {
+@@ -146,6 +159,7 @@ public final class SpawnerCreature {
+ 
+             }
+         }
++        return mobsSpawned; // Paper
+     }
+ 
+     @Nullable
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index 649159653..9bfd7328d 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -21,6 +21,9 @@ import java.io.BufferedWriter;
+ import java.io.IOException;
+ import java.io.Writer;
+ import java.nio.file.Files;
++import java.util.ArrayList; // Paper
++import java.util.EnumMap; // Paper
++import java.util.HashMap; // Paper
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Map;
+@@ -70,6 +73,7 @@ public class WorldServer extends World {
+     private boolean ticking;
+     @Nullable
+     private final MobSpawnerTrader mobSpawnerTrader;
++    private Map<UUID, Map<EnumCreatureType, Integer>> playerMobTypeMap; // Paper
+ 
+     // CraftBukkit start
+     private int tickPosition;
+@@ -846,6 +850,7 @@ public class WorldServer extends World {
+     }
+ 
+     public Object2IntMap<EnumCreatureType> l() {
++        List<Entity> filteredEntities = new ArrayList<>(); // Paper
+         Object2IntMap<EnumCreatureType> object2intmap = new Object2IntOpenHashMap();
+         ObjectIterator objectiterator = this.entitiesById.values().iterator();
+ 
+@@ -870,14 +875,48 @@ public class WorldServer extends World {
+                         entity.spawnReason == CreatureSpawnEvent.SpawnReason.CHUNK_GEN)) {
+                     continue;
+                 }
++                filteredEntities.add(entity);
+                 // Paper end
+                 object2intmap.mergeInt(enumcreaturetype, 1, Integer::sum);
+             }
+         }
+ 
++        // Paper start
++        if(this.paperConfig.perPlayerMobSpawns) {
++            this.playerMobTypeMap = new HashMap<>(this.players.size());
++            updatePlayerMobTypeMap(filteredEntities);
++        }
++        // Paper end
+         return object2intmap;
+     }
+ 
++    // Paper start
++    public void updatePlayerMobTypeMap(List<Entity> entities) {
++        if (this.playerMobTypeMap == null) { return; }
```

Comment:
the map needs to be cleared at this point, you're accumulating data

Reply:
Unless I am mistaken, `playerMobTypeMap` is cleared every tick, or at least each time the spawnmob method is called.

Reply:
yeah sorry it is